pupepat/test/pupil-synth/src/source.py

```diff
@@ -0,0 +1,157 @@
+                                                            counts=self.counts, x=self.x, y=self.y)
+
+    @property
+    def decal(self):
+        shape = (2*self.outer_radius + 1, 2*self.outer_radius + 1)
+        _decal = np.zeros(shape, dtype=np.float_)
+
+        ci, cj = self.half_width, self.half_width  # center of decal
+        ix, jx = np.meshgrid(np.arange(shape[0]), np.arange(shape[1]))
+
+        # calculate dist of all points (ix, jx) to the decal center (ci, cj)
+        dist = np.sqrt((ix-ci)**2 + (jx-cj)**2)
+        _decal[np.where(dist < self.outer_radius)] = self.counts  # fill the circle with counts
+
+        # remake center and dist array for offset inner_radius
+        ci -= self.center_vector[0]
+        cj -= self.center_vector[1]
+        dist = np.sqrt((ix-ci)**2 + (jx-cj)**2)
+        _decal[np.where(dist < self.inner_radius)] = 0.0  # clear the counts within the inner_radius
+        return _decal
+
+
+class EllipticalPupil(Source):
+    def __init__(self,
+                 inner_a, inner_b, inner_phi,
+                 outer_a, outer_b, outer_phi,
+                 counts,
+                 center_vector=(0, 0),
+                 x=None, y=None):
+        super().__init__(x, y)
+        self.inner_a = inner_a
+        self.inner_b = inner_b
+        self.inner_phi = inner_phi
+        self.outer_a = outer_a
+        self.outer_b = outer_b
+        self.outer_phi = outer_phi
+        self.counts = counts
+        # center_vector is the vector from inner to outer center
+        self.center_vector = center_vector
+
+        # TODO: possible validations:
+        # A > B, so we can make that assumption (swap them, don't throw)
+        # outer_b > inner_a
+        # outer_b > inner_a + max(center_vector) (inner doesn't intersect outer)
+        # outer_b > inner_a at worst when inner_phi and outer_phi orthogonal
+        # etc
+
+        self.half_width = max(outer_a, outer_b)  # major semi-axis
+
+    def __str__(self):
+        return 'EllipticalPupil(inner_a={inner_a:0.2f}, inner_b={inner_b:0.2f}, inner_phi={inner_phi:0.2f}, ' \
+               'outer_a={outer_a:0.2f}, outer_b={outer_b:0.2f}, outer_phi={outer_phi:0.2f}, counts={counts:0g} ' \
+               'x={x:}, y={y:})'.format(inner_a=self.inner_a, inner_b=self.inner_b, inner_phi=self.inner_phi,
+                                        outer_a=self.outer_a, outer_b=self.outer_b, outer_phi=self.outer_phi,
+                                        counts=self.counts, x=self.x, y=self.y)
+
+    @property
+    def decal(self):
+        shape = (2*self.half_width + 1, 2*self.half_width + 1)
+        _decal = np.zeros(shape, dtype=np.float_)
+
+        ci, cj = self.half_width, self.half_width  # center of decal
+        ix, jx = np.meshgrid(np.arange(shape[0]), np.arange(shape[1]))
+
+        # calculate dist of all points (ix, jx) to the decal center (ci, cj)
+        dist = (((np.cos(self.outer_phi)*(ix-ci) +
```

Comment:
This looks like a rotation matrix. It might help to split into x and y first. Also I always get the sign on the sin terms backward from this. We should make sure that has a unit test.

Reply:
extracted method. unit test in progress....

Reply:
changed to conventional form...